logger/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+use std::{
+    fs::File,
+    io::{self, Write},
+    sync::Mutex,
+    time::Instant,
+};
+
+use chrono::Utc;
+use once_cell::sync::OnceCell;
+
+static LOGGER: OnceCell<Logger> = OnceCell::new();
+
+struct LoggerImpl {
+    pub sink: Box<dyn Write + Send>,
+    pub start_instant: Instant,
+}
+
+impl LoggerImpl {
+    fn new(kind: LogKind) -> Self {
+        let start_instant = Instant::now();
+        match kind {
+            LogKind::STDOUT => Self {
+                sink: Box::new(io::stdout()),
+                start_instant,
+            },
+            LogKind::FILE => {
+                let now = Utc::now();
+                let path = format!("/tmp/clementine-{}-{}.log", now.date(), now.time());
+                Self {
+                    sink: Box::new(File::create(path).unwrap()),
+                    start_instant,
+                }
+            }
+        }
+    }
+
+    fn log<T>(&mut self, data: T)
+    where
+        T: std::fmt::Display,
+    {
+        let now = self.start_instant.elapsed();
+        let seconds = now.as_secs();
+        let hours = seconds / 3600;
+        let minutes = (seconds / 60) % 60;
+        let seconds = seconds % 60;
+        let miliseconds = now.subsec_millis();
+
+        let _ = writeln!(
+            self.sink,
+            "[{:02}:{:02}:{:02}.{:03}] {}",
+            hours, minutes, seconds, miliseconds, data
+        );
+    }
+}
+
+/// LogKind represents the kind of logging are available now.
+#[derive(PartialEq, Eq)]
+pub enum LogKind {
+    /// It logs to console, the default choice.
+    STDOUT,
+
+    /// It logs on a file in /tmp/clementine-<timestamp>.log
+    FILE,
+}
+
+/// Logger
+struct Logger {
+    pub inner_impl: Mutex<LoggerImpl>,
+}
+
+impl Default for Logger {
+    fn default() -> Self {
+        Self {
+            inner_impl: Mutex::new(LoggerImpl::new(LogKind::STDOUT)),
+        }
+    }
+}
+
+impl Logger {
+    fn new(kind: LogKind) -> Self {
+        Self {
+            inner_impl: Mutex::new(LoggerImpl::new(kind)),
+        }
+    }
+
+    fn log<T>(&self, data: T)
+    where
+        T: std::fmt::Display,
+    {
+        if let Ok(ref mut inner) = self.inner_impl.lock() {
+            inner.log(data);
+        }
+    }
+}
+
+pub fn init_logger(kind: LogKind) {
+    _ = LOGGER.set(Logger::new(kind));
+}
+
+pub fn log<T>(data: T)
+where
+    T: std::fmt::Display,
+{
+    LOGGER.get().map_or((), |logger| logger.log(data));
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Looks good.
Quick question, to you think to modify the tests to make them pass on windows? (It’s no longer my case, but maybe someone runs it on windows)

Reply:
Sure

Reply:
Just for information, what was the problem?

Reply:
Creating a file in tmp directory using `:` cause a strange error on windows, now the file is called `clementine-<timestamp>.log` and before I have used a date with time like `12:00`
Another observation is that in unix system we use `/tmp` for temporary data instead under windows is `C:/users/<name>/appdata/temp`